source/common/http/conn_manager_impl.cc

```diff
@@ -756,6 +767,25 @@ void ConnectionManagerImpl::ActiveStream::resetIdleTimer() {
 
 void ConnectionManagerImpl::ActiveStream::onIdleTimeout() {
   connection_manager_.stats_.named_.downstream_rq_idle_timeout_.inc();
+
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.override_request_timeout_by_gateway_timeout")) {
+    filter_manager_.streamInfo().setResponseFlag(StreamInfo::ResponseFlag::StreamIdleTimeout);
+    sendLocalReply(maybeRequestTimeoutCode(filter_manager_.remoteDecodeComplete()),
+                   "stream timeout", nullptr, absl::nullopt,
+                   StreamInfo::ResponseCodeDetails::get().StreamIdleTimeout);
+    return;
+  }
+
+  // There are 2 issues in the blow code. First, `responseHeaders().has_value()` is not the best
+  // predicate. `remoteDecodeComplete()` is preferable. Second, `sendLocalReply()` smartly ends the
+  // stream if any response was pushed to decoder and explicitly `endStream()` is not required.
+  //
+  // The above code is expected to resolve both. The original code here before it is fully verified.
```

Comment:
The rest of `sendLocalReply(Http::Code::RequestTimeout` is straight forward to be replaced.

This piece has more branches. However, from high level goal, the below whole can be replaced as other occurrences . 
```
 sendLocalReply(maybeRequestTimeoutCode(filter_manager_.remoteDecodeComplete()) ...
```
@mattklein123 Do the code and explanation make sense?

Reply:
Yeah I think you are right and this is just legacy before various fixes, but I would definitely want @alyssawilk to confirm.

Reply:
yeah I think it is legacy and like switching to sendLocalReply, just wanted it behind the guard.